File: src/main/java/net/lomeli/lomlib/lib/Vector3.java
package net.lomeli.lomlib.lib;

import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.BlockPos;
import net.minecraft.util.Vec3;

import net.lomeli.lomlib.util.MathHelper;

public class Vector3 {
    private double x, y, z;

    public Vector3(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Vector3(Vector3 vec) {
        this(vec.x, vec.y, vec.z);
    }

    public Vector3 set(Vector3 vec) {
        x = vec.x;
        y = vec.y;
        z = vec.z;
        return this;
    }

    public Vector3 set(double d, double d1, double d2) {
        x = d;
        y = d1;
        z = d2;
        return this;
    }

    public Vec3 toVec3() {
        return new Vec3(x, y, z);
    }

    public BlockPos toBlockPos() {
        return new BlockPos(MathHelper.floor(x), MathHelper.floor(y), MathHelper.floor(z));
    }

    public Vector3 add(double d, double d1, double d2) {
        return new Vector3(x + d, y + d1, z + d2);
    }

    public Vector3 add(Vector3 vec) {
        return add(vec.x, vec.y, vec.z);
    }

    public Vector3 add(double d) {
        return add(d, d, d);
    }

    public Vector3 sub(double d, double d1, double d2) {
        return new Vector3(x - d, y - d1, z - d2);
    }

    public Vector3 sub(Vector3 vec) {
        return sub(vec.x, vec.y, vec.z);
    }

    public Vector3 sub(double d) {
        return sub(d, d, d);
    }

    public Vector3 mul(double d, double d1, double d2) {
        return new Vector3(x * d, y * d1, z * d2);
    }

    public Vector3 mul(Vector3 vec) {
        return mul(vec.x, vec.y, vec.z);
    }

    public Vector3 mul(double d) {
        return mul(d, d, d);
    }

    public Vector3 div(double d, double d1, double d2) {
        return new Vector3(x / d, y / d1, z / d2);
    }

    public Vector3 div(Vector3 vec) {
        return div(vec.x, vec.y, vec.z);
    }

    public Vector3 div(double d) {
        return div(d, d, d);
    }

    public Vector3 eulerAngles() {
        return new Vector3(Math.toDegrees(Math.atan2(x, z)), Math.toDegrees(-Math.atan2(y, Math.hypot(z, x))), 0);
    }

    public double magnitudeSquared() {
        return x * x + y * y + z * z;
    }

    public double magnitude() {
        return Math.sqrt(magnitudeSquared());
    }

    public double distance(Vector3 v) {
        return this.clone().sub(v).magnitude();
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public double getZ() {
        return z;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Vector3) {
            Vector3 vec = (Vector3) obj;
            return x == vec.x && y == vec.y && z == vec.z;
        }
        return false;
    }

    public double angle(Vector3 vec) {
        return Math.acos(clone().normalize().dotProduct(vec.clone().normalize()));
    }

    public Vector3 normalize() {
        double d = magnitudeSquared();
        if (d != 0)
            mul(1 / d);
        return this;
    }

    public double dotProduct(Vector3 vec) {
        double d = vec.x * x + vec.y * y + vec.z * z;

        if (d > 1 && d < 1.00001)
            d = 1;
        else if (d < -1 && d > -1.00001)
            d = -1;
        return d;
    }

    public double dotProduct(double d, double d1, double d2) {
        return d * x + d1 * y + d2 * z;
    }

    public Vector3 crossProduct(Vector3 vec) {
        double d = y * vec.z - z * vec.y;
        double d1 = z * vec.x - x * vec.z;
        double d2 = x * vec.y - y * vec.x;
        x = d;
        y = d1;
        z = d2;
        return this;
    }

    @Override
    public String toString() {
        return "Vector3[" + x + "," + y + "," + z + "]";
    }

    @Override
    public int hashCode() {
        long hx = Double.doubleToLongBits(x);
        long hy = Double.doubleToLongBits(y);
        long hz = Double.doubleToLongBits(z);
        Long hash = (hx ^ (hx >>> 32));
        hash = 31 * hash + hy ^ (hy >>> 32);
        hash = 31 * hash + hz ^ (hz >>> 32);
        return hash.intValue();
    }

    public Vector3 clone() {
        return new Vector3(x, y, z);
    }

    public static Vector3 fromEntity(Entity e) {
        return new Vector3(e.posX, e.posY, e.posZ);
    }

    public static Vector3 fromEntityCenter(Entity e) {
        return new Vector3(e.posX, e.posY - e.getYOffset() + e.height / 2, e.posZ);
    }

    public static Vector3 fromPlayer(EntityPlayer player, float partialTicks) {
        return new Vector3(player.prevPosX + (player.posX - player.prevPosX) * partialTicks, player.prevPosY + (player.posY - player.prevPosY) * partialTicks, player.prevPosZ + (player.posZ - player.prevPosZ) * partialTicks);
    }

    public static Vector3 fromTile(TileEntity tile) {
        return fromBlockPos(tile.getPos());
    }

    public static Vector3 fromTileCenter(TileEntity tile) {
        return fromTile(tile).add(0.5);
    }

    public static Vector3 fromBlockPos(BlockPos pos) {
        return new Vector3(pos.getX(), pos.getY(), pos.getZ());
    }

    public static Vector3 fromBlockCenter(BlockPos pos) {
        return fromBlockPos(pos).add(0.5);
    }

    public static Vector3 fromAxes(double[] da) {
        return new Vector3(da[2], da[0], da[1]);
    }
}

File: src/main/kotlin/net/lomeli/lomlib/lib/Vector3.kt
package net.lomeli.lomlib.lib

import net.lomeli.lomlib.util.MathUtil
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3d

class Vector3 {
    private var x: Double = 0.toDouble()
    private var y: Double = 0.toDouble()
    private var z: Double = 0.toDouble()

    constructor()

    constructor(x: Double, y: Double, z: Double) {
        this.x = x
        this.y = y
        this.z = z
    }

    constructor(vec: Vector3) : this(vec.x, vec.y, vec.z)

    constructor(vec3: Vec3d) : this(vec3.xCoord, vec3.yCoord, vec3.zCoord)

    constructor(entity: Entity) : this(entity.posX, entity.posY, entity.posZ)

    operator fun set(d: Double, d1: Double, d2: Double): Vector3 {
        x = d
        y = d1
        z = d2
        return this
    }

    fun toVec3(): Vec3d = Vec3d(x, y, z)

    fun toBlockPos(): BlockPos = BlockPos(MathUtil.floor(x), MathUtil.floor(y), MathUtil.floor(z))

    operator fun plus(vec: Vector3): Vector3 {
        x += vec.x
        y += vec.y
        z += vec.z
        return this
    }

    operator fun plus(d: Double): Vector3 = this + Vector3(d, d, d)

    operator fun minus(vec: Vector3): Vector3 {
        x -= vec.x
        y -= vec.y
        z -= vec.z
        return this
    }

    operator fun minus(d: Double): Vector3 = this - Vector3(d, d, d)

    operator fun times(vec: Vector3): Vector3 {
        x *= vec.x
        y *= vec.y
        z *= vec.z
        return this
    }

    operator fun times(d: Double): Vector3 = this * Vector3(d, d, d)

    operator fun div(vec: Vector3): Vector3 {
        x /= vec.x
        y /= vec.y
        z /= vec.z
        return this
    }

    operator fun div(d: Double): Vector3 = this / Vector3(d, d, d)

    operator fun mod(vec: Vector3): Vector3 {
        x %= vec.x
        y %= vec.y
        z %= vec.z
        return this
    }

    operator fun mod(d: Double): Vector3 = this % Vector3(d, d, d)

    fun eulerAngles(): Vector3 = Vector3(Math.toDegrees(Math.atan2(x, z)), Math.toDegrees(-Math.atan2(y, Math.hypot(z, x))), 0.toDouble())

    fun magnitudeSquared(): Double = x * x + y * y + z * z

    fun magnitude(): Double = Math.sqrt(magnitudeSquared())

    fun distance(v: Vector3): Double = (this.clone() - v).magnitude()

    fun getX(): Double = x

    fun getY(): Double = y

    fun getZ(): Double = z

    override fun equals(obj: Any?): Boolean = if (obj is Vector3) equals(obj) else false

    fun equals(x0: Double, y0: Double, z0: Double): Boolean = x == x0 && y == y0 && z == z0

    fun equals(vec: Vector3): Boolean = equals(vec.x, vec.y, vec.z)

    fun angle(vec: Vector3): Double = Math.acos(clone().normalize().dotProduct(vec.clone().normalize()))

    fun normalize(): Vector3 {
        val d = magnitudeSquared()
        if (d != 0.0)
            times(1 / d)
        return this
    }

    fun dotProduct(vec: Vector3): Double {
        var d = vec.x * x + vec.y * y + vec.z * z

        if (d > 1 && d < 1.00001)
            d = 1.0
        else if (d < -1 && d > -1.00001)
            d = -1.0
        return d
    }

    fun dotProduct(d: Double, d1: Double, d2: Double): Double = d * x + d1 * y + d2 * z

    fun crossProduct(vec: Vector3): Vector3 {
        val d = y * vec.z - z * vec.y
        val d1 = z * vec.x - x * vec.z
        val d2 = x * vec.y - y * vec.x
        x = d
        y = d1
        z = d2
        return this
    }

    override fun toString(): String = "Vector3[$x,$y,$z]"

    override fun hashCode(): Int {
        val hx = java.lang.Double.doubleToLongBits(x)
        val hy = java.lang.Double.doubleToLongBits(y)
        val hz = java.lang.Double.doubleToLongBits(z)
        var hash: Long? = hx xor hx.ushr(32)
        hash = 31 * hash!! + hy xor hy.ushr(32)
        hash = 31 * hash + hz xor hz.ushr(32)
        return hash.toInt()
    }

    fun clone(): Vector3 = Vector3(x, y, z)

    fun fromEntity(e: Entity): Vector3 = Vector3(e.posX, e.posY, e.posZ)

    fun fromEntityCenter(e: Entity): Vector3 = Vector3(e.posX, e.posY - e.yOffset + e.height / 2, e.posZ)

    fun fromPlayer(player: EntityPlayer, partialTicks: Float): Vector3 = Vector3(player.prevPosX + (player.posX - player.prevPosX) * partialTicks, player.prevPosY + (player.posY - player.prevPosY) * partialTicks, player.prevPosZ + (player.posZ - player.prevPosZ) * partialTicks)

    fun fromTile(tile: TileEntity): Vector3 = fromBlockPos(tile.pos)

    fun fromTileCenter(tile: TileEntity): Vector3 = fromTile(tile) + 0.5

    fun fromBlockPos(pos: BlockPos): Vector3 = Vector3(pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble())

    fun fromBlockCenter(pos: BlockPos): Vector3 = fromBlockPos(pos) + 0.5

    fun fromAxes(da: DoubleArray): Vector3 = Vector3(da[2], da[0], da[1])
}